src/storage/hummock_sdk/src/slice_transform.rs

```diff
@@ -0,0 +1,436 @@
+        &full_key[0..0]
+    }
+}
+
+// SchemaSliceTransform build from table_catalog and transform a full_key to prefix for
+// prefix_bloom_filter
+pub struct SchemaSliceTransform {
+    // Each stateful operator has its own read pattern, partly using `Prefix Scan`.
+    // `PrefixKeyLength` can be decoded through its `DataType` and `OrderType` which obtained from
+    // `TableCatalog`. `read_pattern_prefix_column` means the count of column to decode prefix from
+    // `StorageKey`
+    read_pattern_prefix_column: u32,
+    deserializer: OrderedRowDeserializer,
+    // TODO:need some bench test for same prefix case like join (if we need a prefix_cache for same
+    // prefix_key)
+}
+
+impl SliceTransform for SchemaSliceTransform {
+    fn transform<'a>(&self, full_key: &'a [u8]) -> &'a [u8] {
+        if full_key.len() < TABLE_PREFIX_LEN + VIRTUAL_NODE_SIZE {
+            panic!(
+                "full_key {:?} len {} not match the schema",
+                full_key,
+                full_key.len()
+            );
+        }
+
+        let (_table_prefix, key) = full_key.split_at(TABLE_PREFIX_LEN);
+        let (_vnode_prefix, pk) = key.split_at(VIRTUAL_NODE_SIZE);
+
+        // if the key with table_id deserializer fail from schema, that shoud panic here for early
+        // detection
+        let pk_prefix_len = self
+            .deserializer
+            .deserialize_prefix_len_with_column_indices(
+                pk,
+                (0..self.read_pattern_prefix_column as usize).collect(),
+            )
+            .unwrap();
+
+        let prefix_len = TABLE_PREFIX_LEN + VIRTUAL_NODE_SIZE + pk_prefix_len;
+        &full_key[0..prefix_len]
+    }
+}
+
+impl SchemaSliceTransform {
+    fn new(table_catalog: &Table) -> Self {
+        assert_ne!(0, table_catalog.read_pattern_prefix_column);
+
+        // column_index in pk
+        let pk_indices: Vec<usize> = table_catalog
+            .order_key
+            .iter()
+            .map(|col_order| col_order.index as usize)
+            .collect();
+
+        let data_types = pk_indices
+            .iter()
+            .map(|column_idx| &table_catalog.columns[*column_idx])
+            .map(|col| ColumnDesc::from(col.column_desc.as_ref().unwrap()).data_type)
+            .collect();
+
+        let order_types: Vec<OrderType> = table_catalog
+            .order_key
+            .iter()
+            .map(|col_order| {
+                OrderType::from_prost(
+                    &risingwave_pb::plan_common::OrderType::from_i32(col_order.order_type).unwrap(),
+                )
+            })
+            .collect();
+
+        Self {
+            read_pattern_prefix_column: table_catalog.read_pattern_prefix_column,
+            deserializer: OrderedRowDeserializer::new(data_types, order_types),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct MultiSliceTransForm {
+    id_to_slice_transform: HashMap<u32, Arc<dyn SliceTransform>>,
```

Comment:
We may define a `SliceTransformImpl` with enum dispatch in the future since this might be dynamically dispatched frequently.